# 🧠 Day 3 – SQL via Python: NYC School Data Exploration
In this notebook, you'll connect to a PostgreSQL database and execute SQL queries to explore NYC school data.

Database Tables:

high_school_directory – School names, locations, types, programs

school_demographics – Enrollment data, ELL, FRPL, disabilities, etc.

school_safety_report – Reported incidents by type and location.

## 🔌 Step 1: Import Libraries

In [54]:
import pandas as pd
import psycopg2

## 🔐 Step 2: Connect to the Database

In [55]:
# DB connection setup using hardcoded credentials (for onboarding only)
conn = psycopg2.connect(
    dbname="neondb",
    user="neondb_owner",
    password="npg_CeS9fJg2azZD",
    host="ep-falling-glitter-a5m0j5gk-pooler.us-east-2.aws.neon.tech",
    port="5432",
    sslmode="require"
)
cur = conn.cursor()

## 🔍 Step 3: Run a Test Query

In [56]:
query = "SELECT * FROM nyc_schools.high_school_directory LIMIT 5;"
df = pd.read_sql(query, conn)
df.head()

/var/folders/92/fx7kkcl54x92rqxpxpp474500000gn/T/ipykernel_3287/1731849884.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, conn)


dbn                                        school_name    borough  \
0  27Q260          Frederick Douglass Academy VI High School     Queens   
1  21K559        Life Academy High School for Film and Music   Brooklyn   
2  16K393     Frederick Douglass Academy IV Secondary School   Brooklyn   
3  08X305                               Pablo Neruda Academy      Bronx   
4  03M485  Fiorello H. LaGuardia High School of Music & A...  Manhattan   

  building_code  phone_number    fax_number  grade_span_min  grade_span_max  \
0          Q465  718-471-2154  718-471-2890             9.0              12   
1          K400  718-333-7750  718-333-7775             9.0              12   
2          K026  718-574-2820  718-574-2821             9.0              12   
3          X450  718-824-1682  718-824-1663             9.0              12   
4          M485  212-496-0700  212-724-5748             9.0              12   

  expgrade_span_min expgrade_span_max  ... number_programs  \
0              None              None  ...               1   
1              None              None  ...               1   
2              None              None  ...               1   
3              None              None  ...               1   
4              None              None  ...               6   

                                          Location 1 Community Board  \
0  {'latitude': '40.601989336', 'longitude': '-73...              14   
1  {'latitude': '40.593593811', 'longitude': '-73...              13   
2  {'latitude': '40.692133704', 'longitude': '-73...               3   
3  {'latitude': '40.822303765', 'longitude': '-73...               9   
4  {'latitude': '40.773670507', 'longitude': '-73...               7   

  Council District Census Tract Zip Codes Community Districts  \
0               31       100802     20529                  51   
1               47          306     17616                  21   
2               36          291     18181                  69   
3               18           16     11611                  58   
4                6          151     12420                  20   

  Borough Boundaries City Council Districts Police Precincts  
0                  3                     47               59  
1                  2                     45               35  
2                  2                     49               52  
3                  5                     31               26  
4                  4                     19               12  

[5 rows x 105 columns]

## ✅ How many schools are there in each borough?

In [57]:
# Example: Count schools by borough
query = """
SELECT borough, COUNT(*) AS school_count
FROM nyc_schools.high_school_directory
GROUP BY borough;
"""
df_result = pd.read_sql(query, conn)
df_result

/var/folders/92/fx7kkcl54x92rqxpxpp474500000gn/T/ipykernel_3287/3450537001.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_result = pd.read_sql(query, conn)


borough  school_count
0       Brooklyn           121
1         Queens            80
2  Staten Island            10
3      Manhattan           106
4          Bronx           118

In [58]:
# Joining Tables: High School Directory + School Demographics
query = """
SELECT * FROM nyc_schools.high_school_directory
LEFT JOIN nyc_schools.school_demographics
ON nyc_schools.high_school_directory.dbn = nyc_schools.school_demographics.dbn;
"""
df_dir_demo = pd.read_sql(query, conn)
df_dir_demo

/var/folders/92/fx7kkcl54x92rqxpxpp474500000gn/T/ipykernel_3287/196753222.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_dir_demo = pd.read_sql(query, conn)


dbn                                        school_name    borough  \
0    01M292      Henry Street School for International Studies  Manhattan   
1    01M292      Henry Street School for International Studies  Manhattan   
2    01M292      Henry Street School for International Studies  Manhattan   
3    01M292      Henry Street School for International Studies  Manhattan   
4    01M292      Henry Street School for International Studies  Manhattan   
..      ...                                                ...        ...   
463  14K685            El Puente Academy for Peace and Justice   Brooklyn   
464  22K555                           Brooklyn College Academy   Brooklyn   
465  14K478  The High School for Enterprise, Business and T...   Brooklyn   
466  24Q296             Pan American International High School     Queens   
467  14K614        Young Women's Leadership School of Brooklyn   Brooklyn   

    building_code  phone_number    fax_number  grade_span_min  grade_span_max  \
0            M056  212-406-9411  212-406-9417             6.0              12   
1            M056  212-406-9411  212-406-9417             6.0              12   
2            M056  212-406-9411  212-406-9417             6.0              12   
3            M056  212-406-9411  212-406-9417             6.0              12   
4            M056  212-406-9411  212-406-9417             6.0              12   
..            ...           ...           ...             ...             ...   
463          K778  718-387-1125  718-387-4229             9.0              12   
464          K917  718-853-6184  718-853-6356             9.0              12   
465          K450  718-387-2800  718-387-2748             9.0              12   
466          Q744  718-271-3602  718-271-4041             9.0              12   
467          K147  718-387-5641           N/A             6.0              12   

     expgrade_span_min  expgrade_span_max  ... black_num black_per  \
0                  NaN                NaN  ...     106.0      36.1   
1                  NaN                NaN  ...     137.0      31.6   
2                  NaN                NaN  ...     158.0      30.7   
3                  NaN                NaN  ...     138.0      29.4   
4                  NaN                NaN  ...     141.0      27.6   
..                 ...                ...  ...       ...       ...   
463                NaN                NaN  ...       NaN       NaN   
464                NaN                NaN  ...       NaN       NaN   
465                NaN                NaN  ...       NaN       NaN   
466                NaN                NaN  ...       NaN       NaN   
467                NaN                NaN  ...       NaN       NaN   

    hispanic_num hispanic_per white_num white_per male_num male_per  \
0          133.0         45.2      10.0       3.4    160.0     54.4   
1          208.0         47.9      14.0       3.2    241.0     55.5   
2          272.0         52.8      12.0       2.3    281.0     54.6   
3          264.0         56.2      14.0       3.0    264.0     56.2   
4          290.0         56.8      16.0       3.1    297.0     58.1   
..           ...          ...       ...       ...      ...      ...   
463          NaN          NaN       NaN       NaN      NaN      NaN   
464          NaN          NaN       NaN       NaN      NaN      NaN   
465          NaN          NaN       NaN       NaN      NaN      NaN   
466          NaN          NaN       NaN       NaN      NaN      NaN   
467          NaN          NaN       NaN       NaN      NaN      NaN   

    female_num female_per  
0        134.0       45.6  
1        193.0       44.5  
2        234.0       45.4  
3        206.0       43.8  
4        214.0       41.9  
..         ...        ...  
463        NaN        NaN  
464        NaN        NaN  
465        NaN        NaN  
466        NaN        NaN  
467        NaN        NaN  

[468 rows x 143 columns]

In [59]:
df_dir_demo['ell_percent'].isna().sum()

np.int64(428)

## What is the average % of ELL per borough?

In [60]:
# Count the average % of ELL students per borough
df_ell_count = df_dir_demo.groupby('borough')['ell_percent'].agg(['mean', 'count']).sort_values('mean', ascending=False)
df_ell_count

mean  count
borough                     
Manhattan      7.5725     40
Bronx             NaN      0
Brooklyn          NaN      0
Queens            NaN      0
Staten Island     NaN      0

The English Language Learners are only in Manhattan and the average of its percentage is 7.5 

## Top 3 schools in each borough with the highest percentage of special education students

Using the data from the school demographics and high school directory

In [90]:
# Show all rows that have duplicate DBNs (including both original and duplicate)
duplicates = df_dir_demo[df_dir_demo.duplicated(subset=['dbn'], keep=False)]
duplicates[['dbn', 'sped_percent']]

dbn     dbn  sped_percent
0   01M292  01M292          19.0
1   01M292  01M292          22.4
2   01M292  01M292          24.1
3   01M292  01M292          25.1
4   01M292  01M292          21.9
5   01M292  01M292          23.7
6   01M292  01M292          24.9
7   01M448  01M448          15.7
8   01M448  01M448          15.6
9   01M448  01M448          19.0
10  01M448  01M448          21.7
11  01M448  01M448          23.1
12  01M448  01M448          22.7
13  01M448  01M448          21.8
14  01M450  01M450          24.8
15  01M450  01M450          25.1
16  01M450  01M450          24.5
17  01M450  01M450          26.7
18  01M450  01M450          28.8
19  01M450  01M450          27.7
20  01M450  01M450          26.4
21  01M509  01M509          24.4
22  01M509  01M509          23.3
23  01M509  01M509          21.1
24  01M509  01M509          18.4
25  01M509  01M509          21.3
26  01M509  01M509          21.1
27  01M509  01M509          25.9
29  01M539  01M539           4.9
30  01M539  01M539           4.6
31  01M539  01M539           4.0
32  01M539  01M539           3.8
33  01M539  01M539           3.2
34  01M539  01M539           2.8
35  01M539  01M539           2.7
36  01M696  01M696           0.2
37  01M696  01M696           0.2
38  01M696  01M696           0.2
39  01M696  01M696           0.5

In [ ]:
# Remove duplicate columns if present
df_dir_demo_fixed = df_dir_demo.loc[:, ~df_dir_demo.columns.duplicated()]

# Remove duplicates by keeping the average sped_percent for each DBN
df_dir_demo_clean = df_dir_demo_fixed.groupby('dbn').agg({
    'school_name': 'first',  
    'borough': 'first',      
    'sped_percent': 'mean',  # Calculate average of sped_percent
})
df_dir_demo_clean

school_name    borough  \
dbn                                                                    
01M292      Henry Street School for International Studies  Manhattan   
01M448                University Neighborhood High School  Manhattan   
01M450                         East Side Community School  Manhattan   
01M509                            Marta Valle High School  Manhattan   
01M539  New Explorations into Science, Technology and ...  Manhattan   
...                                                   ...        ...   
32K545      EBC High School for Public Service - Bushwick   Brooklyn   
32K549                 Bushwick School for Social Justice   Brooklyn   
32K552                          Academy of Urban Planning   Brooklyn   
32K554               All City Leadership Secondary School   Brooklyn   
32K556  Bushwick Leaders High School for Academic Exce...   Brooklyn   

        sped_percent  
dbn                   
01M292     23.014286  
01M448     19.942857  
01M450     26.285714  
01M509     22.214286  
01M539      3.714286  
...              ...  
32K545           NaN  
32K549           NaN  
32K552           NaN  
32K554           NaN  
32K556           NaN  

[435 rows x 3 columns]

In [103]:
# Get top 3 schools per borough with highest special education percentages
top_3_sped = (
    df_dir_demo_clean
    .reset_index()  # 'dbn' as a column again
    .groupby('borough')
    .apply(lambda x: x.nlargest(3, 'sped_percent')[['dbn', 'school_name', 'sped_percent']])
)
top_3_sped

/var/folders/92/fx7kkcl54x92rqxpxpp474500000gn/T/ipykernel_3287/671743175.py:6: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(lambda x: x.nlargest(3, 'sped_percent')[['dbn', 'school_name', 'sped_percent']])


dbn                                        school_name  \
borough                                                                        
Bronx         106  07X221    South Bronx Preparatory: A College Board School   
              107  07X223  M.S. 223 The Laboratory School of Finance and ...   
              108  07X259  H.E.R.O. High (Health, Education, and Research...   
Brooklyn      224  13K265  Dr. Susan S. McKinney Secondary School of the ...   
              225  13K350        Urban Assembly High School of Music and Art   
              226  13K412  Brooklyn Community High School of Communicatio...   
Manhattan     2    01M450                         East Side Community School   
              0    01M292      Henry Street School for International Studies   
              3    01M509                            Marta Valle High School   
Queens        338  24Q236      International High School for Health Sciences   
              339  24Q264                  Academy of Finance and Enterprise   
              340  24Q267               High School of Applied Communication   
Staten Island 418  31R047          CSI High School for International Studies   
              419  31R064        Gaynor McCown Expeditionary Learning School   
              420  31R080                     The Michael J. Petrides School   

                   sped_percent  
borough                          
Bronx         106           NaN  
              107           NaN  
              108           NaN  
Brooklyn      224           NaN  
              225           NaN  
              226           NaN  
Manhattan     2       26.285714  
              0       23.014286  
              3       22.214286  
Queens        338           NaN  
              339           NaN  
              340           NaN  
Staten Island 418           NaN  
              419           NaN  
              420           NaN

## 🧠 Insights

1) There are 121 schools in Brooklyn, 80 in Queens, 10 in Staaten Island, 106 in Manhattan and 118 schoöls in Bronx. 

2) The English Language Learners are only in Manhattan and the average of its percentage is 7.5 

3) Top 3 schools in  Manhattan with the highest percentage of special education students are the schools with 'dbn' : 01M450, 01M292 and 01M509	